Update and rename PEP.rst to PEP.md (#1)
zooba/python-api-pep: PEP.md

CPython API Design Guidelines

# Abstract

This document is intended to be a set of guiding principles for development of the current CPython API. Future additions and enhancements to the CPython API should follow, or at least be influenced by, the principles described here. At a minimum, any new or modified APIs should be able to be categorised according to the terminology defined here, even if exceptions have to be made.

> # This document is NOT ...

Reviewer: FWIW, I really like the format of this and the following section:   `This document is NOT ... (though ...).` :)

This document is NOT a design of a completely new API (though the design of a hypothetical new API should be guided by this document).

This document is NOT documentation of the current API (though the principles in this document should become more obvious in the current API over time).

This document is NOT a set of binding rules in the same sense as PEP 7 and PEP 8 (though designs should be tested against it and new exceptions should be rare).

This document is NOT permission to make backwards-incompatible modifications to the current API (though backwards-incompatible modifications should still be made where justified).

# This document IS ...

> This document is language agnostic, and applies to both Python and C APIs (though language-specific references are used for the benefit of readers familiar with CPython's implementation).

Reviewer: The application to Python APIs is a big deal but might be lost on folks at first.  This is okay as long as the point is treated well later in the PEP.

This document is meant to be referenced in discussions (though preferably by concept rather than section number).

> This document is supposed to cause arguments about how to classify existing API members (though with shared definitions it should be possible to reach agreement).

Reviewer: Pure gold! :)

> # Definitions

Reviewer: The changes you've already made to the definitions from my earlier feedback are much appreciated. :)  The text is definitely clearer now, especially "rings" and "layers".

A common understanding of certain terms is necessary to talking about any API, and so this section defines the terms we use to discuss the CPython API. This section has two goals: to clarify existing common terminology, and to introduce new terminology. Terms are presented in a logical order, rather than alphabetically.

## Existing terms

**Application**: A program that can be launched directly. Compare and contrast with *extension*. CPython is normally considered an application when used via the `python` or `python3` command.

> **Extension**: A program that cannot be launched directly but must be loaded by an application. Python modules, native or otherwise, are considered extenions. When embedded into another application, CPython is considered an extension.

Reviewer: ```suggestion
**Extension**: A program that cannot be launched directly but must be loaded by an application. Python modules, native or otherwise, are considered extensions. When embedded into another application, CPython is considered an extension.
```

> **Extension**: A program that cannot be launched directly but must be loaded by an application. Python modules, native or otherwise, are considered extenions. When embedded into another application, CPython is considered an extension.

Reviewer: "native" is slightly ambiguous here, but that's probably fine.  You clarify well enough just a little further down.  Just in case, here's how I would make it more clear:

```suggestion
**Extension**: A program that cannot be launched directly but must be loaded by an application. Python modules, whether "native" or pure-Python, are considered extenions. When embedded into another application, CPython is considered an extension.
```

**Target Runtime**: The combination of operating system and externally-provided code that an application is compiled to use. Referred to as "target runtime" to distinguish from "Python runtime". For CPython, the target runtime is usually the processor, operating system and C runtime implementation.

**Native extension**: A subset of all extensions that are compiled to the same target runtime as the application they integrate with. CPython native extensions are compiled to native code that uses the CPython ABI. For contrast, Python source and bytecode files are *not* considered native extensions.

> **API**: Application Programming Interface. The set of interactions defined by an application to allow extensions to extend, control, and interact with the application. An API typically defines functions, classes and values as abstract interfaces. CPython has one API that applies for all scenarios in all contexts, though practical scenarios only use a subset of this API. (Referring to a subset of the API as "the <x> API" is generally acceptable.)

Reviewer: FWIW, I find the reference to "classes" to be slightly jarring, since CPython's C-API only has C declarations (functions, variables, structs, etc.).

> **API**: Application Programming Interface. The set of interactions defined by an application to allow extensions to extend, control, and interact with the application. An API typically defines functions, classes and values as abstract interfaces. CPython has one API that applies for all scenarios in all contexts, though practical scenarios only use a subset of this API. (Referring to a subset of the API as "the <x> API" is generally acceptable.)

Reviewer: Where you say `An API typically defines functions, classes and values as abstract interfaces.`, it makes "API" a very concrete thing in my mind (for whatever reason).  The work "declarations" helps mitigate that for me:

```suggestion
**API**: Application Programming Interface. The set of interactions defined by an application to allow extensions to extend, control, and interact with the application. An API typically provides declarations of functions, classes and values as abstract interfaces. CPython has one API that applies for all scenarios in all contexts, though practical scenarios only use a subset of this API. (Referring to a subset of the API as "the <x> API" is generally acceptable.)
```

Reviewer: The concept of opaque types is pretty important for the API, but I don't see it mentioned.  I'm on the fence of if it's important enough to include in the definition for "API", but it should at least be pointed at some point in the PEP.

> **ABI**: Application Binary Interface. The implementation of an API such that its interactions can be realized by a digital computer. Typically includes memory layouts and binary representations, and is a function of the target platform and the build tools used to compile CPython. A particular build of CPython may have a different ABI (but identical API) to another build, which will impact compatibility with native extension builds but not sources.

Reviewer: FWIW, I liked the implication (in the previous version) that there's a sort of ABI for bytecode files.  However, I expect that isn't really a part of the focus of this PEP. :)

Author: .pyc files may be the most interesting application of the API/ABI definitions... my gut feel:
* "python myfile.pyc" is the API
* "import myfile" (resolves to .pyc) is the API
* "open('myfile.pyc', 'rb').read()" (and equivalent) is the ABI

**Stdlib**: Standard library. The set of components developed by the core team that build upon the Python language to provide building blocks and pre-written functionality.

## New terms

These terms are introduced briefly here and described in much greater detail below.

> **API rings**: A way of dividing an API into subsets to specify stability guarantees. CPython extensions will care about rings and will allow themselves access to a particular ring to choose between deeper integration or long-term stability. Allowing access to one ring implies access to all rings outside of that one, which are "more stable" or "more accessible" (imagine a castle containing the most private API subset with multiple ringed walls around it, each ring providing "more public" APIs). Rings are orthogonal to layers.

Reviewer: That castle analogy is a great addition.  It crystallizes the meaning.  Plus, you get bonus points for the subtle Monty Python reference. :)

> **API rings**: A way of dividing an API into subsets to specify stability guarantees. CPython extensions will care about rings and will allow themselves access to a particular ring to choose between deeper integration or long-term stability. Allowing access to one ring implies access to all rings outside of that one, which are "more stable" or "more accessible" (imagine a castle containing the most private API subset with multiple ringed walls around it, each ring providing "more public" APIs). Rings are orthogonal to layers.

Reviewer: FWIW, while most readers will understand what you mean by "stability" in the context of APIs, it may be confusing for readers that think of stability in terms of likelihood to fail at runtime (rather than at build-time).  I don't have any recommendations for mitigating that.  Sorry. :)

> **API layers**: A way of dividing an API into subsets to specify interdependencies and coupling. Applications that embed CPython, and the CPython implementation itself, care about layers. Applications choose to adopt or implement a particular layer, which requires that all "lower" API layers must be made available, either by inclusion or implementation (imagine a building with foundations, where each layer of the building is only feasible because of the layers below it). Layers are orthogonal to rings.

Reviewer: Likewise with that building example.  Nice work.

> **API member**: One element of an API, such as a function, a class, or an accesible data value. Every user of an API will use it by accessing individual members. API members should belong to exactly one API ring and exactly one API layer.

Reviewer: (disregarding the relationship of classes to the Python C-API...)

I guess I only think of functions and data values as fundamental API members, since that's the real interaction.  Classes/objects seem more like de-facto namespaces.  If classes count then I'd expect other compound data types, like struct, to be included.

That said, you clearly aren't enumerating an exhaustive list. :)  I only wanted to point out that the inclusion of "class" but the exclusion of "struct" jumped out at me (as a reader).

# Quick Overview

> For context as you continue reading, these are the API **rings** provided by CPython:

Reviewer: I was going to suggest pointing out the correlation between the proposed rings and the existing divisions in the C-API (i.e. the different include directories).  However, doing so at this point in the document might muddle the concepts too much.  It may make sense to point of that correlation later in the document, when the concepts are already more firmly established in readers' minds (but only if you can avoid the same muddling effect there).

> For context as you continue reading, these are the API **rings** provided by CPython:

Reviewer: Order here is important, but already relatively clear.  That said, in clarifying the significance of the order you can reinforce a key relationship:

```suggestion
For context as you continue reading, these are the API **rings** provided by CPython (in order from greatest stability to least stability):
```

> * Python ring (equivalent of the Python language)

Reviewer: FWIW, at first it's a little hard to grasp what the rings really represent here.  It makes more sense after reading the next section (about stability guarantees), so the text here is more meaningful only on the second pass through the doc.  I'm not sure that there's much to be done about this though.  It's probably okay.  I just wanted to point out that I was hung up on this for a few seconds.

Perhaps the following wording would help?

```suggestion
* Python ring (API equivalent of just the Python language)
```

or 

```suggestion
* Python ring (API equivalent of just the Python language capability and semantics)
```

Also, "API" is less clear than "CPython's C-API", but is more correct, so perhaps provide the latter as a concrete example?

```suggestion
* Python ring (API equivalent of just the Python language, e.g. in CPython's C-API)
```

> * CPython ring (CPython-specific APIs)

Reviewer: Likewise for clarity, perhaps the following wording would help:

```suggestion
* CPython ring (APIs specific to the CPython implementation)
```

> * CPython ring (CPython-specific APIs)

Reviewer: If the "Python" ring relates strictly to the Python language and "CPython" to the implementation then it may make sense to include a ring in between for general Python-related API that is not specific to CPython.

The catch is that the stability guarantee of that ring would coincide with that of "Python" or "CPython", thus weakening the focus of "ring" on stability guarantees.  I suppose this could be mitigated by allowing such helpers to be part of the "Python" ring.  The downside there is it would make it harder for people to decide what belongs in the "Python" ring; "exactly Python-the-language" is a much easier metric than implementation-agnostic Python stuff.

Author: I'm considering calling it:

`CPython's Python ring (CPython spellings for generic Python APIs)`

But ultimately, it's probably best to just say that `PyObject_Call` is a CPython API, and the Python API spelling is `callable_object()` :)

> * Internal ring (intended for internal use only)

Reviewer: This is fine as-is, but maybe only because I'm well-versed in the distinction between public and internal API with CPython. :)

> These are the API **layers** provided by CPython:

Reviewer: The order matters here.  Though it's relatively clear from the names and descriptions, it may be worth pointing out the order (especially since you use the reverse order in the `Availability Guarantees Quick Overview` section below):

```suggestion
These are the API **layers** provided by CPython (in order from most coupling to least coupling):
```

It might make sense to be likewise clear in that later section.

* Optional stdlib layer (dependencies that must be explicitly required)
* Required stdlib layer (dependencies that can be assumed)
* Platform interaction layer (ability to interact with the platform and user)
* Core layer ("pure" mode with no dependency on target runtime)

> * Platform adaption layer (isolates core layer from the target runtime)

Reviewer: FWIW, I think what you have here is sufficient for this point in the document, though these all demand further explanation (which you give in subsequent sections).

Reviewer: Also, the names are good because, if you refer to just the name later in the PEP, few readers will have to come back to this section for a definition. :)

(Reminder that this document does not reflect the current state of CPython, but is both aspirational and defining terms for the purposes of discussion. This is not a proposal to remove anything from the standard distribution!)

> ## Stability Guarantees Quick Overview

Reviewer: It may pay off to subtly remind readers right away that "Stability Guarantees" is the province of rings.  Once way is to refer to them in the heading:

```suggestion
## Stability Guarantees Quick Overview (API Rings)
```

Another would be to point it out explicitly at the beginning of the section:

```suggestion
## Stability Guarantees Quick Overview

(Reminder: stability guarantees are the responsibility of API rings)
```

Author: I'm actually switching this up to make stability guarantees for both rings and layers, while rings will exclusively cover generality (e.g. "does this apply to PyPy too?") and layers will cover coupling (e.g. "is X allowed to depend on Y") and availability (e.g. "does X being present mean that Y will be present too").

For context as you continue reading, these are half-sentence summaries of the stability guarantees made by CPython for each ring:

* Python ring APIs change only at `x` version number changes
* CPython ring APIs change only at `x.y` version number changes

> * Internal ring APIs may change at any release

Reviewer: ```suggestion
* Internal ring APIs may change at any release, including bug-fix
```

> ## Availability Guarantees Quick Overview

Reviewer: Likewise with reminding readers about layers:

```suggestion
## Availability Guarantees Quick Overview (API Layers)
```

or

```suggestion
## Availability Guarantees Quick Overview

(Reminder: availability guarantees are the responsibility of API layers)
```

For context as you continue reading, these are half-sentence summaries of the availability guarantees made by CPython for each layer:

* Platform adaptation layer will always be present, but cannot be introspected (it is a "black box")
* Core layer will always be available
* Platform interaction layer may be present, but can only be accessed using APIs in the core layer
* Required stdlib layer is available when any of the optional stdlib layer is present

> * Optional stdlib layer is made up of multiple components, any of which being present implies that the entire required stdlib layer and core layer are present

Reviewer: If I'm understanding right then the "optional stdlib" layer is effectively composed of multiple "sibling" layers that each group component-oriented API subsets.  If that is correct then here are some questions:

* can sibling layers overlap?
* do they have to be non-interdependent?  (otherwise the "optional stdlib" layers must be composed vertically in addition to horizontally)
* would "layer components" be the preferred term (you call them "components" here)?
* would it make sense to apply the same concept of sibling layers to any of the lower layers?
* if so, how would that tie into sibling layers up the stack (e.g. how would sibling layers in "required stdlib" relate to those in "optional stdlib")?

The downside to allowing "sibling" layers below "optional stdlib" (i.e. the most outer layer) is the conceptual complexity it adds.  If "sibling" layers only exist in the outermost layer then things stay relatively simple.  So I doubt it's a good idea to introduce them in any other layer.

Author: I like the terminology of "sibling" components, because yeah, that's basically what they are. I feel like I might need to draw an actual graphic with boxes to show this part though... still thinking about it

> 

Reviewer: <end review part 1>
